### DISCUSSION TITLE
How to use ImageFromDockerfile.withDockerfileFromBuilder as compatible substitute image?

### DISCUSSION THREAD

**Discussion by nylund** [testcontainers/testcontainers-java#5104] [Q&A] [answered]
Please provide a way to allow customization of an image when pulling it through a proxy. For example this snippet does not work if ELASTICSEARCH_IMAGE uses an in-house/local docker proxy:

```    
ImageFromDockerfile customImage = new ImageFromDockerfile()
    .withDockerfileFromBuilder(builder ->
        builder
        .from(ELASTICSEARCH_IMAGE)
        .run("bin/elasticsearch-plugin", "install", "analysis-kuromoji")
        .build());
```

Ends up in exception:

```
java.lang.IllegalStateException: Failed to verify that image 'localhost/testcontainers/shreetkrpuzjpwzg' is a compatible substitute for 'docker.elastic.co/elasticsearch/elasticsearch'. This generally means that you are trying to use an image that Testcontainers has not been designed to use. If this is deliberate, and if you are confident that the image is compatible, you should declare compatibility in code using the `asCompatibleSubstituteFor` method. For example:
   DockerImageName myImage = DockerImageName.parse("localhost/testcontainers/shreetkrpuzjpwzg").asCompatibleSubstituteFor("docker.elastic.co/elasticsearch/elasticsearch");
and then use `myImage` instead.
```

Is there a way to turn off this type of validation?

Version:  1.16.3

**Comment by kiview** [comment]
Did you try the code as provided by the exception message?
```java
DockerImageName myImage = DockerImageName.parse("localhost/testcontainers/shreetkrpuzjpwzg")
  .asCompatibleSubstituteFor("docker.elastic.co/elasticsearch/elasticsearch");
```

This exception happens when you start the Elasticsearch container, it has nothing to do with pulling through a proxy.

**Comment by nylund** [comment]
I did that for another case but I cannot wrestle the API to use DockerImageName instances in this case. The builder.from() method only takes a String.

**Comment by nylund** [comment]
And note that 'localhost/testcontainers/shreetkrpuzjpwzg' is not a name from our code but that seems to be auto generated by something inside TestContainers. ELASTICSEARCH_IMAGE is in our case set to "in-house-proxy/elasticsearch/elasticsearch:6.8.8".

**Comment by kiview** [comment, marked as answer]
Can you try something along the lines of:
```java
DockerImageName myImage = DockerImageName.parse(customImage.get())
  .asCompatibleSubstituteFor("docker.elastic.co/elasticsearch/elasticsearch");
```

**Comment by nylund** [comment]
`resolve()` is protected so I can't call that one.

**Comment by bsideup** [comment]
.get() is the method

**Comment by nylund** [comment]
Great, this works! Thanks so much for the fast feedback.

FYI - I am asking around internally if our company can become sponsors for Testcontainers. No promises but I'll see what I can do.

**Comment by kiview** [comment]
Great we could solve it for your use case 🙂 
I will try to close and convert this issue into a GitHub Discussion question, let's see if this works as expected. 

Also, we really appreciate advocating for Testconainers in your organization and helping others adopt it for their integration testing needs.
